test/common/upstream/load_balancer_benchmark.cc

```diff
@@ -139,6 +195,56 @@ void computeHitStats(benchmark::State& state,
   state.counters["relative_stddev_hits"] = (stddev / mean);
 }
 
+void BM_RoundRobinChooseHost(benchmark::State& state) {
+  const uint64_t keys_to_simulate = 100000;
+
+  for (auto _ : state) {
+    // Do not time the creation.
+    state.PauseTiming();
+    const uint64_t num_hosts = state.range(0);
+    const uint64_t weighted_subset_percent = state.range(1);
+    const uint64_t weight = state.range(2);
+    RoundRobinTester tester(num_hosts, weighted_subset_percent, weight);
+    tester.initialize();
+
+    std::unordered_map<std::string, uint64_t> hit_counter;
+    TestLoadBalancerContext context;
+    state.ResumeTiming();
+
+    for (uint64_t i = 0; i < keys_to_simulate; i++) {
+      // Note: To a certain extent this is benchmarking the performance of
+      // std::unordered_map.
+      // TODO(antoniovicente): Consider removing the hit counter to avoid polluting the benchmark
+      //                       numbers.
```

Comment:
We could do two passes, one with, one without stats. and only take the performance from the latter. I'm not sure how much this is a map benchmark though, do you have figures with/without?

Reply:
Output with "hit_counter[tester.lb_->chooseHost(&context)->address()->asString()] += 1;"
time  benchmark
 37 ms BM_RoundRobinChooseHost/500/0/1
 42 ms BM_RoundRobinChooseHost/2500/0/1
 59 ms BM_RoundRobinChooseHost/10000/0/1
150 ms BM_RoundRobinChooseHost/50000/0/1
 48 ms BM_RoundRobinChooseHost/500/5/2
 55 ms BM_RoundRobinChooseHost/2500/5/2
 74 ms BM_RoundRobinChooseHost/10000/5/2
172 ms BM_RoundRobinChooseHost/50000/5/2
 48 ms BM_RoundRobinChooseHost/500/50/75
 56 ms BM_RoundRobinChooseHost/2500/50/75
 73 ms BM_RoundRobinChooseHost/10000/50/75
161 ms BM_RoundRobinChooseHost/50000/50/75
 49 ms BM_RoundRobinChooseHost/500/100/75
 55 ms BM_RoundRobinChooseHost/2500/100/75
 75 ms BM_RoundRobinChooseHost/10000/100/75
169 ms BM_RoundRobinChooseHost/50000/100/75

Output with "tester.lb_->chooseHost(&context);"
time  benchmark
 21 ms BM_RoundRobinChooseHost/500/0/1
 24 ms BM_RoundRobinChooseHost/2500/0/1
 35 ms BM_RoundRobinChooseHost/10000/0/1
 96 ms BM_RoundRobinChooseHost/50000/0/1
 38 ms BM_RoundRobinChooseHost/500/5/2
 42 ms BM_RoundRobinChooseHost/2500/5/2
 55 ms BM_RoundRobinChooseHost/10000/5/2
124 ms BM_RoundRobinChooseHost/50000/5/2
 39 ms BM_RoundRobinChooseHost/500/50/75
 43 ms BM_RoundRobinChooseHost/2500/50/75
 57 ms BM_RoundRobinChooseHost/10000/50/75
124 ms BM_RoundRobinChooseHost/50000/50/75
 38 ms BM_RoundRobinChooseHost/500/100/75
 43 ms BM_RoundRobinChooseHost/2500/100/75
 56 ms BM_RoundRobinChooseHost/10000/100/75
120 ms BM_RoundRobinChooseHost/50000/100/75

For completeness, I tried the following:
void BM_RoundRobinChooseHost(benchmark::State& state) {
  const uint64_t keys_to_simulate = 100000;

  for (auto _ : state) {
    // Do not time the creation.                                                                                                                                                                                                                                                    
    state.PauseTiming();
    const uint64_t num_hosts = state.range(0);
    const uint64_t weighted_subset_percent = state.range(1);
    const uint64_t weight = state.range(2);
    RoundRobinTester tester(num_hosts, weighted_subset_percent, weight);
    tester.initialize();

    TestLoadBalancerContext context;
    state.ResumeTiming();

    for (uint64_t i = 0; i < keys_to_simulate; i++) {
      tester.lb_->chooseHost(&context);
    }

    state.PauseTiming();
    // Compute stats as a separate pass in order to avoid polluting the benchmark.                                                                                                                                                                                                
    std::unordered_map<std::string, uint64_t> hit_counter;
    for (uint64_t i = 0; i < keys_to_simulate; i++) {
      hit_counter[tester.lb_->chooseHost(&context)->address()->asString()] += 1;
    }
    computeHitStats(state, hit_counter);

    // Exclude destructors from timing.                                                                                                                                                                                                                                             
  }
}
The results were closer to the output with
"hit_counter[tester.lb_->chooseHost(&context)->address()->asString()] += 1;", see below:
44 ms BM_RoundRobinChooseHost/500/0/1
43 ms BM_RoundRobinChooseHost/2500/0/1
58 ms BM_RoundRobinChooseHost/10000/0/1
133 ms BM_RoundRobinChooseHost/50000/0/1
61 ms BM_RoundRobinChooseHost/500/5/2
69 ms BM_RoundRobinChooseHost/2500/5/2
93 ms BM_RoundRobinChooseHost/10000/5/2
168 ms BM_RoundRobinChooseHost/50000/5/2
59 ms BM_RoundRobinChooseHost/500/50/75
68 ms BM_RoundRobinChooseHost/2500/50/75
84 ms BM_RoundRobinChooseHost/10000/50/75
168 ms BM_RoundRobinChooseHost/50000/50/75
59 ms BM_RoundRobinChooseHost/500/100/75
70 ms BM_RoundRobinChooseHost/2500/100/75
89 ms BM_RoundRobinChooseHost/10000/100/75
183 ms BM_RoundRobinChooseHost/50000/100/75

It seems strange that work done while timing is paused affects the result.

Reply:
I'm curious if the cost comes from generating and hashing the address string, rather than the hash map itself. In any case, I agree, the stats bookkeeping is high.